File: VB/Reporting_runtime-table-creation-best-practices-iterative-approach-e1356/XtraReport1.vb
Imports System.Drawing
Imports System.Drawing.Printing
Imports DevExpress.XtraReports.UI
' ...

Namespace XRTable_RuntimeCreation
    Partial Public Class XtraReport1
        Inherits XtraReport

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub XtraReport1_BeforePrint(ByVal sender As Object, ByVal e As PrintEventArgs) Handles Me.BeforePrint
            Me.Detail.Controls.Add(CreateXRTable())
        End Sub

        Public Function CreateXRTable() As XRTable
            Dim cellsInRow As Integer = 3
            Dim rowsCount As Integer = 3
            Dim rowHeight As Single = 25F

            Dim table As New XRTable()
            table.Borders = DevExpress.XtraPrinting.BorderSide.All
            table.BeginInit()

            For i As Integer = 0 To rowsCount - 1
                Dim row As New XRTableRow()
                row.HeightF = rowHeight
                For j As Integer = 0 To cellsInRow - 1
                    Dim cell As New XRTableCell()
                    row.Cells.Add(cell)
                Next j
                table.Rows.Add(row)
            Next i

            AddHandler table.BeforePrint, AddressOf table_BeforePrint
            table.AdjustSize()
            table.EndInit()
            Return table
        End Function

        ' The following code makes the table span to the entire page width.
        Private Sub table_BeforePrint(ByVal sender As Object, ByVal e As PrintEventArgs)
            Dim table As XRTable = (DirectCast(sender, XRTable))
            table.LocationF = New DevExpress.Utils.PointFloat(0F, 0F)
            table.WidthF = Me.PageWidth - Me.Margins.Left - Me.Margins.Right
        End Sub
    End Class
End Namespace

File: CS/Reporting_runtime-table-creation-best-practices-iterative-approach-e1356/XtraReport1.cs
using System.Drawing;
using System.Drawing.Printing;
using DevExpress.XtraReports.UI;
// ...

namespace XRTable_RuntimeCreation {
    public partial class XtraReport1 : XtraReport {

        public XtraReport1() {
            InitializeComponent();
        }

        private void XtraReport1_BeforePrint(object sender, PrintEventArgs e) {
            this.Detail.Controls.Add(CreateXRTable());
        }

        public XRTable CreateXRTable() {
            int cellsInRow = 3;
            int rowsCount = 3;
            float rowHeight = 25f;

            XRTable table = new XRTable();
            table.Borders = DevExpress.XtraPrinting.BorderSide.All;
            table.BeginInit();

            for (int i = 0; i < rowsCount; i++) {
                XRTableRow row = new XRTableRow();
                row.HeightF = rowHeight;
                for (int j = 0; j < cellsInRow; j++) {
                    XRTableCell cell = new XRTableCell();
                    row.Cells.Add(cell);
                }
                table.Rows.Add(row);
            }

            table.BeforePrint += new PrintEventHandler(table_BeforePrint);
            table.AdjustSize();
            table.EndInit();
            return table;
        }

        // The following code makes the table span to the entire page width.
        void table_BeforePrint(object sender, PrintEventArgs e) {
            XRTable table = ((XRTable)sender);
            table.LocationF = new DevExpress.Utils.PointFloat(0F, 0F);
            table.WidthF = this.PageWidth - this.Margins.Left - this.Margins.Right;
        }
    }
}
